[Sema] Improve error for inout on conversion

beccadax opened the pull request:
When a user attempts to pass a value requiring an implicit conversion as an `inout` argument (usually a subclass instead of its superclass, or a conforming type instead of its protocol existential), the diagnostic we produce is not very clear:

```swift
func useSuper(_: inout Super) {}
var sub = Sub()
useSuper(&sub)
// error: cannot pass immutable value as inout argument: implicit conversion
//     from 'Sub' to 'Super' requires a temporary
```

This PR changes the error message:
```swift
// error: cannot pass implicitly converted value of type 'Sub' to inout argument:
//     use a temporary variable of exact type 'Super' instead
```

It leaves the existing diagnostic in place for a number of related errors involving implicit conversions (e.g. with operators, on left-hand side of assignment, etc.). I don't think that these cases should actually occur—operators apparently reject solutions with implicit conversions much earlier, and most of the others seem like solutions we'd never generate—but if they can, the old message is better than nothing.

It'd be nice to offer one or both of these fixes:
```swift
// note: make inout parameter generic if 'useSuper' cannot change its type
func useSuper<T: Super>(_: inout T) {}
```
```swift
// note: insert a temporary of type 'Super'
var temporary = sub as Super
useSuper(&temporary)
sub = temporary as! Sub
```

But the former would require us to diagnose this error higher in the AST, probably at the `CallExpr`, while the latter seems to really stretch what a fix-it can currently do.

Resolves [SR-8155](https://bugs.swift.org/browse/SR-8155). See also [SR-8148](https://bugs.swift.org/browse/SR-8148), which was closed by the user.

Tagging Joe because he griped about this diagnostic in a FIXME, and David because he reported the issue.

brentdax commented:
Let's see if the CI bot recognizes me yet...

@swift-ci Please smoke test

jckarter commented:
@jrose-apple usually has good feedback about error message language. I'm concerned that the suggestion in the new message doesn't really address the user's problem when they run into this situation—even if they make a local temporary with the supertype, they probably want to do something with the new value, likely as its original more specific type, and they're still going to by stymied by the type system.

jrose-apple commented:
I personally like something like the second message you suggested better,

> inout argument could be set to a value which cannot be converted to type 'String': use a temporary variable of type 'TextOutputStream' instead

although I might simplify it to "a value with a type other than 'String'". I like this version better because people may not think of it on their own, and it'll help them decide whether the suggested fix is actually correct. The version that's in the PR now is still an improvement, but to me it's just a better phrasing of the "what" and "how" without quite getting the "why".

Grammar nitpick: that should be a semicolon in either version of the message, since the second clause is not subordinate to the first.

brentdax commented:
@jrose-apple I've added two commits. The first changes the wording of the diagnostic—I like the more explanatory tone too. (And the semicolon had occurred to me, but I wasn't sure we were using them in other diagnostics.)

The second commit adds a fix-it which offers to change the type of the original variable:
```swift
SR-8155.swift:7:10: error: inout argument could be set to a value with a type other than 'Sub'; use a value declared as type 'Super' instead
useSuper(&sub)
         ^~~~
SR-8155.swift:6:8: note: change variable type to 'Super' if it doesn't need to be declared as 'Sub'
var sub = Sub()
       ^
       : Super
```
It can correct an existing type or insert a new one. I'm not sure this fix-it is a good idea because changing the type won't always produce working code, but I thought I'd implement it and see what you think.

brentdax commented:
@swift-ci Please smoke test

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -767,11 +828,51 @@ static void diagnoseSubElementFailure(Expr *destExpr,
+          .highlight(ICE->getSourceRange());
+        
+        if (auto *DRE = dyn_cast<DeclRefExpr>(LE->getSubExpr())) {
+          if (auto * VD = dyn_cast_or_null<VarDecl>(DRE->getDecl())) {
```
It's worth checking that this is a local variable; you wouldn't want to offer the fix-it for a global or a property.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -767,11 +828,51 @@ static void diagnoseSubElementFailure(Expr *destExpr,
+                scratch += ": ";
+              }
+              bool typeAppended = false;
+              if (auto *P = VD->getParentPattern()) {
```
You already have this, above.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -767,11 +828,51 @@ static void diagnoseSubElementFailure(Expr *destExpr,
+              }
+              else if (auto *PD = dyn_cast<ParamDecl>(VD)) {
+                // These have typeRanges but not a ParentPattern. Fortunately they
+                // don't destructure either, so we can just use the whole neededType.
```
…which makes this `if` mis-attached, perhaps? Definitely need to add tests.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -654,6 +654,67 @@ static bool isLoadedLValue(Expr *expr) {
+      return true;
+    }
+    
+    bool visitVarPattern(VarPattern *P) { return visit(P->getSubPattern()); }
```
I think you can mark all these `case`-only patterns as `llvm_unreachable` instead.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -654,6 +654,67 @@ static bool isLoadedLValue(Expr *expr) {
+    
+    bool visitVarPattern(VarPattern *P) { return visit(P->getSubPattern()); }
+    bool visitEnumElementPattern(EnumElementPattern *P) { return visit(P->getSubPattern()); }
+    bool visitTypedPattern(TypedPattern *P) { return visit(P->getSubPattern()); }
```
Uh, hang on, what about `let (i: Int, j: Double) = (0, 0)`?

brentdax replied:
Tuple shuffle strikes again! `let (i: Int, j: Double) = (0, 0)` doesn't do what you think it does:

      (pattern_binding_decl range=[<stdin>:1:1 - line:1:32]
        (pattern_tuple type='(i: Int, j: Int)' names=i,j
          (pattern_named type='Int' 'Int')
          (pattern_named type='Int' 'Double'))

jrose-apple replied:
Ugh, right. TypedPattern is `as Int`, not `: Int`.

jrose-apple reviewed (commented):
I think trying to add a type to a tuple pattern is too easy to get wrong (see comment counter-example), so you could just limit the fix-it to simple cases where the pattern is a single variable already. The fix-it is on a note so it might be okay even though it could break other code. @jckarter, what do you think?

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@jrose-apple Sorry, busy couple of days. This will now ignore patterns involving destructuring, though it will still handle variables declared like `let a = 1, b = 2`. It will *mostly* only suggest the fix for local variables, but that was excluding variables declared in top-level code, so I made an exception for that. Also just polished things up in general and added a horde of tests.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -654,65 +654,68 @@ static bool isLoadedLValue(Expr *expr) {
+  // variables in the same file to be local variables.)
+  auto VDC = VD->getDeclContext();
+  bool isLocalVar = VDC->isLocalContext();
+  if (VDC->isModuleScopeContext()) {
```
Minor optimization: check `isLocalVar` first.

jrose-apple reviewed (commented):
Seems reasonable to me at this point! One more test you could add:

`var foo, bar: IsPrintable1`

…but I don't think it matters too much what happens for this one.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -684,18 +684,39 @@ static void fixItChangeInoutArgType(const Expr * arg,
+    if (!PBD) return false;
+    
+    auto nextIndex = PBD->getPatternEntryIndexForVarDecl(VD) + 1;
+    // If VD is last, there's nothing we could be sharing a type with.
```
…What about the previous one? The fix-it you have in the example will change the type of both variables. Again, I don't think that's unreasonable, but in that case it should be the same in either direction.

brentdax replied:
There's a reason I reverted most of this commit and kept only the structure of the test. :^)

brentdax replied:
I was planning to wait and squash-merge this, but I think I'll just squash it immediately to simplify the history.

jrose-apple replied:
Yep, sorry, GitHub showed me the in-between state and immediately after I saw that you dropped it.

beccadax committed:
[Sema] Improve error for inout on conversion

The diagnostic when a user attempts to pass an argument requiring an implicit conversion (e.g. declared as a subclass, type conforming to protocol, etc.) to an inout parameter seems to be confusing for users—see e.g. SR-8155, SR-8148. This change replaces it with a more specific diagnostic which clearly suggests a solution. It also includes a fix-it suggesting the user change the type of the original variable if it’s a local with a simple type signature.

brentdax commented:
I think the Linux failures on the last test were unrelated…

@swift-ci please smoke test

brentdax commented:
@jrose-apple Anything else I should do, or are we good to merge?

jckarter commented:
Looks good to me too.

hartbit commented:
Thanks for working on this @brentdax!
